nesta_ds_utils/networks/build.py

```diff
@@ -78,3 +123,91 @@ def _cooccurrence_counts(sequences):
     return Counter(chain(*combos))
+
+
+def _frequency_counts(all_tokens: List) -> dict:
+    """counts frequency of all terms in corpus
+
+    Args:
+        all_tokens (List): list of all terms in corpus
+
+    Returns:
+        dict: key: token, value: frequency
+    """
+    return Counter(all_tokens)
+
+
+def _jaccard_similarity(edge_weights: dict, all_tokens: list) -> dict:
+    """calculate the jaccard similarity between nodes in the network
+
+    Args:
+        edge_weights (dict): co-occurence counts of the nodes in the network
+        all_tokens (list): list of all tokens in the corpus
+
+    Returns:
+        dict: key: pair of tokens, value: jaccard similarity
+    """
+    token_frequency = _frequency_counts(all_tokens)
+    jaccard_sims = defaultdict(int)
+    for key, val in edge_weights.items():
+        jaccard_sims[key] = val / (
+            (token_frequency[key[0]] + token_frequency[key[1]]) - val
+        )
+    return jaccard_sims
+
+
+def _association_strength(edge_weights: dict, all_tokens: list) -> dict:
+    """calculate the association strength between nodes in the network
+
+    Args:
+        edge_weights (dict): co-occurence counts of the nodes in the network
+        all_tokens (list): list of all tokens in the corpus
+
+    Returns:
+        dict: key: pair of tokens, value: association strength
+    """
+    token_frequency = _frequency_counts(all_tokens)
+    association_strengths = defaultdict(int)
+    for key, val in edge_weights.items():
+        association_strengths[key] = val / (
+            token_frequency[key[0]] * token_frequency[key[1]]
+        )
+    return association_strengths
+
+
+def _cosine_sim(edge_weights: dict, all_tokens: list) -> dict:
+    """calculate the cosine similarity between nodes in the network
+
+    Args:
+        edge_weights (dict): co-occurence counts of the nodes in the network
+        all_tokens (list): list of all tokens in the corpus
+
+    Returns:
+        dict: key: pair of tokens, value: cosine similarity
+    """
+    token_frequency = _frequency_counts(all_tokens)
+    cosine_similarities = defaultdict(int)
+    for key, val in edge_weights.items():
+        cosine_similarities[key] = val / (
+            math.sqrt(token_frequency[key[0]] * token_frequency[key[1]])
+        )
+    return cosine_similarities
+
+
+def _inclusion_index(edge_weights: dict, all_tokens: list) -> dict:
+    """calculate the inclusion index between nodes in the network
+
+    Args:
+        edge_weights (dict): co-occurence counts of the nodes in the network
+        all_tokens (list): list of all tokens in the corpus
+
+    Returns:
+        dict: key: pair of tokens, value: inclusion index
+    """
+    token_frequency = _frequency_counts(all_tokens)
+    inclusion_index = defaultdict(int)
+    for key, val in edge_weights.items():
+        inclusion_index[key] = val / min(
+            token_frequency[key[0]], token_frequency[key[1]]
+        )
+    return inclusion_index
```

Comment:
These look fine to me, but raise a couple of questions:

- How scalable is this approach? Depending on the number of nodes and network structure, the possible number of edges can be huge, in which case vectorised methods would provide a significant speed up.
- Should a user be able to apply these once they have already constructed a graph? This would be more similar to the nx API and would avoid having to reconstruct the graph plus all the weights if you had left one out and then realised you needed it. 
  - In this case, the graph construction might change in that you could build the basic graph with edges first, then calculate and apply edge weights in turn.

Reply:
I'm thinking to leave this as is for now, but possibly create a new feature for another module outside of "build" that takes an existing graph as input and adds features to it, as I don't think that really falls into "build"